### PR TITLE
Robust Examples

### PR DESCRIPTION
**Pull request by Manav1712**
The following pull request contains notebooks which solve the following problems - Robust Facility Location, Robust Portfolio, Robust Knapsack, Robust Lot Sizing, Robust Mean - Variance Portfolio, Robust Production - Inventory and Robust Support Vector Machines.

### CODE REVIEW

**Review by AmitSolomonPrinceton — COMMENTED** [review]
General comments:

1. I recommend having "introduction", "formulation", and "numerical example" subsection for each problem. For example, in the "Robust Facility Location" problem the first markdown is an introduction subsection, and the following subsection is the problem formulation. The block where you give numerical values ("In this example, we define an uncertain parameter $D$ using the Ellipsoidal definition with $\rho=0.3, \mathbf{b}$ as value of $D$ in the deterministic case.") is where the numerical example starts.
2. Please make sure all vectors are denoted as vectors in math form.
3. Please use \left( \right) instead of ( ) in math form.
4. I know you didn't create robust_knapsack and robust_portfolio notebooks, but could you please format them in the same way as the other notebooks? (e.g. same number of decimal points as the other notebooks).
5. Make sure Abbreviations are defined before used.
6. Need to remove the sparsity warnings (I'm okay with removing them manually since the issue is resolved in the develop branch).
7. Please make sure all variables and values are in math mode ($$ in markdown).
8. Make sure that every notebook is self contained. For example, you have to redefine things that have been defined in other notebooks.

File: docs/examples/robust_lot_sizing.ipynb [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "$$\n",
+    "\\begin{array}{ll}\n",
+    "\\text{minimize} & s^TY + v^TX + H \\\\\n",
+    "\\text{subject to} & H_t \\geq h_t[ \\bar{p}^TX_t - d] \\\\\n",
```
[the remark this answers is not in the document]
Comment (Manav1712) [reply docs/examples/robust_lot_sizing.ipynb:38]: This is how the problem was formulated in the paper.

[the remark this answers is not in the document]
Comment (AmitSolomonPrinceton) [reply docs/examples/robust_lot_sizing.ipynb:38]: Our notation uses parentheses.

File: docs/examples/robust_facility.ipynb [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With this information in mind, our optimization problem becomes:\n",
+    "$$\n",
+    "\\begin{array}{ll}\n",
+    "\\text{maximize} & (\\eta \\cdot 1^T - d) \\cdot (x \\cdot D) - c^T (1^T P) - C^T Z - K^T I \\\\\n",
```
[the remark this answers is not in the document]
Comment (Manav1712) [reply docs/examples/robust_facility.ipynb:33]: This is how the problem was formulated in the paper - they used capitals to denote both vectors and scalars. Should I make the scalars lowercase and the vectors uppercase?

File: docs/examples/robust_lot_sizing.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Manav1712) [reply docs/examples/robust_lot_sizing.ipynb:38]: Great, I'll change it accordingly

File: docs/examples/robust_prod_inventory.ipynb [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Taking the example introduced in Ben-Tal et al. (2004) [1], let us consider a single product inventory system which is comprised of a warehouse and $I$ factories. The planning horizon is $T$ periods. At period $t$:\n",
+    "- $d_t$ is the demand of the product that is uncertain\n",
+    "- $v_t$ is the amount of product in the warehouse at the beginning of a period\n",
+    "- $p_t$ is the amount of product to be produced during period $t$ by factory $i$ and is used to satisfy the demand of the period\n",
+    "- $P_t$ is the maximal production capacity of factory $i$ in time period $t$\n",
+    "- $c_t$ is the cost of producing a unit of product at factory $i$ in time period $t$\n",
+    "- $Q_t$ is the maximal cumulative production capacity of factory $i$ in time period $t$\n",
```
[the remark this answers is not in the document]
Comment (Manav1712) [reply docs/examples/robust_prod_inventory.ipynb]: So originally, the variable was formulated as $c_i(t)$, sort of as a function of t. However, Professor Stellato told me to remove the 'function' and write is as a subscript $c_t$. Should I make it $c_it$ instead?

File: docs/examples/robust_facility.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Manav1712) [reply docs/examples/robust_facility.ipynb:33]: For example - p has dimensions (F x T) - I think having it capitalized is fine. Should I change Z, I, d and 1 to lowercase bolded letters?

[the remark this answers is not in the document]
Comment (AmitSolomonPrinceton) [reply docs/examples/robust_facility.ipynb:33]: We should have consistent notation and conventions, regardless of what is used in a specific paper. Please have consistent notation for scalars/vectors/matrices.

File: docs/examples/robust_facility.ipynb [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lets look at each term of the objective function:\n",
+    "\n",
+    "- The expression $ (\\eta \\cdot 1^T - d) \\cdot (x \\cdot D) $ represents the revenue generated from satisfying demand at location  $i$, time period $ \\tau $, and serving demand from location $ j$. It calculates the difference between the revenue $ \\eta $ and the unit demand $ d $, multiplied by the decision variable $x $ and the nominal demand $ D $.\n",
```
[the remark this answers is not in the document]
Comment (AmitSolomonPrinceton) [reply docs/examples/robust_facility.ipynb:49]: If we don't use the indices we should remove them...

File: docs/examples/robust_prod_inventory.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (AmitSolomonPrinceton) [reply docs/examples/robust_prod_inventory.ipynb]: Either c_{i,t}, or remove the $i$ from the description (you cannot have undefined subscripts).

**Comment by bstellato** [comment]
Thanks @Manav1712! Great job. A few comments (I mainly edited the portfolio example to show what I had in mind):
- Make sure problems visualize on jupyterlab (not only on vscode). For example, if you have an empty line in an optimization problem it will not render
- unless strictly needed, I would not include seaborn as a dependency. you can rearrange the for loops to make it easier without it. You can use tab colors and rearrange the font sizes. See how I edited the portfolio notebook.
- Many notebooks start with "The Facility Location Problem (FLP) is a fundamental optimization challenge that focuses on....". That smells like LLM. Try to start with simply "Consider the problem of determining the optimal locations of facilities (e.g., warehouses, plants, or distribution centers)...."
- If you say "we cant to solve the uncertain problem", it needs to show up with the uncertainty set immediately. (forall ...). Also, the "robust counterpart" is the problem where you already get the reformulation without having the "forall" inside the constraints. That's something that the package does automatically so we do not need to write it. See, again, how I updated the portfolio example to simplify it.
- Treat equations like text. This means they all have to end with a comma or a period.
- Avoid writing the descriptin of the uncertainty set in a similar way as the numpy docstrings (e.g., "$A$ : np.array, optional"). Simply follow what I wrote in the portfolio example.

File: docs/examples/robust_prod_inventory.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Irina526) [reply docs/examples/robust_prod_inventory.ipynb]: ### The formulation that you give for this problem also needs to be updated accordingly. The factories "i" must be taken into account somewhere in the formulation. In addition, the last constraint should be for each time step t - it should not be a single constraint.

File: docs/examples/robust_prod_inventory.ipynb [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    "    prob = lropt.RobustProblem(objective, constraints)\n",
+    "    prob.solve()\n",
+    "    return F.value, p.value"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following code calculates the robust optimal level of inventory over time. "
```
Comment (Irina526) [inline docs/examples/robust_prod_inventory.ipynb:212]: For the visualization, can you also generate a random trajectory, for both perturbation = 0.1 and 0.2. The inventory levels for that trajectory should be between the current upper and lower bounds.

File: docs/examples/robust_facility.ipynb [hunk 6]
```diff
@@ -32,12 +32,20 @@
     "With this information in mind, our optimization problem becomes:\n",
     "$$\n",
     "\\begin{array}{ll}\n",
+<<<<<<< HEAD
```
Comment (AmitSolomonPrinceton) [inline docs/examples/robust_facility.ipynb:35]: You need to fix this (also in L42, L48. In short, you need to decide what stream you would like and keep that stream).

**Pull request merged**